### PR TITLE
Task-J49: add enum & validation of episode type

### PR DESCRIPTION
**Pull request by vitalik-prog**

### CODE REVIEW

File: backend/src/data/migrations/20210801114612_add_type_to_episodes.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { Knex } from 'knex';
+
+const TABLE_NAME = 'episodes';
+const COLUMN_NAME = 'type';
+
+export async function up(knex: Knex): Promise<void> {
+  return knex.schema.table(TABLE_NAME, (table) => {
+    table.string(COLUMN_NAME).notNullable();
```
Comment (ghost) [inline backend/src/data/migrations/20210801114612_add_type_to_episodes.ts]:
```suggestion
    table.enum(COLUMN_NAME).notNullable();
```

File: shared/src/validation-schemas/episode/episode.validation-schema.ts [hunk 2]
```diff
@@ -30,6 +30,8 @@ const episode = Joi.object<EpisodeCreatePayload>({
       'number.required': EpisodeValidationMessage.PODCAST_ID_REQUIRE,
       'number.integer': EpisodeValidationMessage.PODCAST_ID_NUMBER_FORMAT,
     }),
+  [EpisodeCreatePayloadKey.TYPE]: Joi.string()
+    .valid(...Object.values(EpisodeTypeKey))
```
Comment (ghost) [inline shared/src/validation-schemas/episode/episode.validation-schema.ts]:
```suggestion
    .valid(Object.values(EpisodeTypeKey)).required(),
```

File: shared/src/common/enums/episode/episode-type-key.enum.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+enum EpisodeTypeKey {
```
Comment (ghost) [inline shared/src/common/enums/episode/episode-type-key.enum.ts]:
```suggestion
enum EpisodeType {
```

File: shared/src/common/types/episode/episode-create-payload.type.ts [hunk 4]
```diff
@@ -4,6 +4,7 @@ type EpisodeCreatePayload = {
   [EpisodeCreatePayloadKey.NAME]: string;
   [EpisodeCreatePayloadKey.USER_ID]: number;
   [EpisodeCreatePayloadKey.PODCAST_ID]: number;
+  [EpisodeCreatePayloadKey.TYPE]: string;
```
Comment (ghost) [inline shared/src/common/types/episode/episode-create-payload.type.ts]:
```suggestion
  [EpisodeCreatePayloadKey.TYPE]: EpisodeType;
```

File: shared/src/common/types/episode/episode.type.ts [hunk 5]
```diff
@@ -5,6 +5,7 @@ type Episode = {
   podcastId: number;
   createdAt: string;
   updatedAt: string;
+  type: string;
```
Comment (ghost) [inline shared/src/common/types/episode/episode.type.ts]:
```suggestion
  type: EpisodeType;
```

File: shared/src/common/enums/episode/episode-type-key.enum.ts (hunk 3 above)
Comment (ghost) [inline shared/src/common/enums/episode/episode-type-key.enum.ts]: rename to `shared/src/common/enums/episode/episode-type.enum.ts`

File: backend/src/data/models/episode/episode.model.ts [hunk 6]
```diff
@@ -8,6 +8,8 @@ class Episode extends Abstract {
 
   [EpisodeDTOKey.PODCAST_ID]: number;
 
+  [EpisodeDTOKey.TYPE]: string;
```
Comment (ghost) [inline backend/src/data/models/episode/episode.model.ts]:
```suggestion
  [EpisodeDTOKey.TYPE]: EpisodeType;
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: shared/src/validation-schemas/episode/episode.validation-schema.ts (hunk 2 above)
Comment (vitalik-prog) [reply shared/src/validation-schemas/episode/episode.validation-schema.ts]: Without this destructuring "...Object.values" application crashes

File: backend/src/data/migrations/20210801114612_add_type_to_episodes.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { Knex } from 'knex';
+
+const TABLE_NAME = 'episodes';
+const COLUMN_NAME = 'type';
+
+export async function up(knex: Knex): Promise<void> {
+  return knex.schema.table(TABLE_NAME, (table) => {
+    table.enu(COLUMN_NAME, ['public', 'private', 'unlisted']).defaultTo('public');
```
Comment (ghost) [inline backend/src/data/migrations/20210801114612_add_type_to_episodes.ts]:
```suggestion
    table.enum(COLUMN_NAME, [EpisodeType.PUBLIC, EpisodeType. PRIVATE, EpisodeType. UNLISTED]).defaultTo(EpisodeType.PUBLIC);
```

File: shared/src/validation-schemas/episode/episode.validation-schema.ts [hunk 8]
```diff
@@ -30,6 +30,12 @@ const episode = Joi.object<EpisodeCreatePayload>({
       'number.required': EpisodeValidationMessage.PODCAST_ID_REQUIRE,
       'number.integer': EpisodeValidationMessage.PODCAST_ID_NUMBER_FORMAT,
     }),
+  [EpisodeCreatePayloadKey.TYPE]: Joi.string()
+    .valid(...Object.values(EpisodeType))
```
Comment (ghost) [inline shared/src/validation-schemas/episode/episode.validation-schema.ts]:
```suggestion
    .valid(Object.values(EpisodeType))
```

File: backend/src/data/migrations/20210801114612_add_type_to_episodes.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { Knex } from 'knex';
+
+const TABLE_NAME = 'episodes';
+const COLUMN_NAME = 'type';
+
```
Comment (ghost) [inline backend/src/data/migrations/20210801114612_add_type_to_episodes.ts:11]:
```tsx
const EpisodeType = {
  PUBLIC: 'public',
  PRIVATE: 'private',
  UNLISTED: 'unlisted',
} as const;
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by qwiz1 — DISMISSED** [review]

**Review by ghost — DISMISSED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**